File: src/game/game_features_debug.cpp
#include "game_config.h"

#include "widget/debug_console.h"
#include "imgui/imgui.h"
#include "imgui/imgui_internal.h"
#include "core/bstring.h"
#include <cctype>

ANK_REGISTER_PROPS_ITERATOR(config_show_game_features_properties);

static bool string_contains_ci(const char* str, const char* substr) {
    if (!substr || !*substr) {
        return true; // Empty filter matches everything
    }
    if (!str || !*str) {
        return false;
    }

    const char* s = str;
    const char* p = substr;
    
    while (*s) {
        const char* s_check = s;
        const char* p_check = p;
        
        while (*s_check && *p_check && 
               std::tolower(static_cast<unsigned char>(*s_check)) == 
               std::tolower(static_cast<unsigned char>(*p_check))) {
            s_check++;
            p_check++;
        }
        
        if (!*p_check) {
            return true; // Found match
        }
        
        s++;
    }
    
    return false;
}

static bool feature_matches_filter(game_features::game_feature* feature, const char* filter) {
    if (!filter || !*filter) {
        return true; // No filter means show all
    }
    
    if (string_contains_ci(feature->name.c_str(), filter)) {
        return true;
    }
    
    if (!feature->text.empty() && string_contains_ci(feature->text.c_str(), filter)) {
        return true;
    }
    
    return false;
}

void config_show_game_features_properties(bool header) {
    if (header) {
        return;
    }

    bool features_open = ImGui::TreeNodeEx("Game Features", ImGuiTreeNodeFlags_None, "Game Features");
    if (!features_open) {
        return;
    }

    static char filter_buffer[256] = "";
    ImGui::Text("Filter:");
    ImGui::SameLine();
    ImGui::SetNextItemWidth(ImGui::GetContentRegionAvail().x - 100);
    ImGui::InputText("##filter", filter_buffer, sizeof(filter_buffer));
    ImGui::SameLine();
    if (ImGui::Button("Clear")) {
        filter_buffer[0] = '\0';
    }

    auto all_features = game_features::all();

    ImGui::BeginTable("GameFeatures", 2, ImGuiTableFlags_BordersOuter | ImGuiTableFlags_Resizable);

    int bool_count = 0;
    int string_count = 0;
    int float_count = 0;
    int vec2i_count = 0;
    int total_count = static_cast<int>(all_features.size());

    for (auto* feature : all_features) {
        switch (feature->type()) {
            case setting_bool:
                bool_count++;
                break;
            case setting_string:
                string_count++;
                break;
            case setting_float:
                float_count++;
                break;
            case setting_vec2i:
                vec2i_count++;
                break;
        }
    }

    game_debug_show_property("total_features", total_count, true);
    game_debug_show_property("bool_features", bool_count, true);
    game_debug_show_property("string_features", string_count, true);
    if (float_count > 0) {
        game_debug_show_property("float_features", float_count, true);
    }
    if (vec2i_count > 0) {
        game_debug_show_property("vec2i_features", vec2i_count, true);
    }

    ImGui::EndTable();

    int filtered_bool_count = 0;
    int filtered_string_count = 0;
    for (auto* feature : all_features) {
        if (!feature_matches_filter(feature, filter_buffer)) {
            continue;
        }
        if (feature->type() == setting_bool) {
            filtered_bool_count++;
        } else if (feature->type() == setting_string) {
            filtered_string_count++;
        }
    }

    bstring256 bool_label;
    bool_label.printf("Bool Features (%d)", filtered_bool_count);
    bool bool_features_open = ImGui::TreeNodeEx("Bool Features", ImGuiTreeNodeFlags_None, "%s", bool_label.c_str());
    if (bool_features_open) {
        ImGui::BeginTable("BoolFeatures", 2, ImGuiTableFlags_BordersOuter | ImGuiTableFlags_Resizable);

        for (auto* feature : all_features) {
            if (feature->type() != setting_bool) {
                continue;
            }

            if (!feature_matches_filter(feature, filter_buffer)) {
                continue;
            }

            bstring256 label;
            label.printf("%s", feature->name.c_str());
            bool current_value = feature->to_bool();
            bool save = current_value;
            game_debug_show_property(label.c_str(), current_value);
            if (save != current_value) {
                feature->set(current_value);
            }
        }

        ImGui::EndTable();
        ImGui::TreePop();
    }

    bstring256 string_label;
    string_label.printf("String Features (%d)", filtered_string_count);
    bool string_features_open = ImGui::TreeNodeEx("String Features", ImGuiTreeNodeFlags_None, "%s", string_label.c_str());
    if (string_features_open) {
        ImGui::BeginTable("StringFeatures", 2, ImGuiTableFlags_BordersOuter | ImGuiTableFlags_Resizable);

        for (auto* feature : all_features) {
            if (feature->type() != setting_string) {
                continue;
            }

            if (!feature_matches_filter(feature, filter_buffer)) {
                continue;
            }

            bstring256 label;
            label.printf("%s", feature->name.c_str());
            xstring current_value = feature->to_string();
            game_debug_show_property(label.c_str(), current_value.c_str());
        }

        ImGui::EndTable();
        ImGui::TreePop();
    }

    int filtered_total = 0;
    for (auto* feature : all_features) {
        if (feature_matches_filter(feature, filter_buffer)) {
            filtered_total++;
        }
    }
    
    bstring256 detailed_label;
    detailed_label.printf("All Features (Detailed) (%d)", filtered_total);
    bool detailed_open = ImGui::TreeNodeEx("All Features (Detailed)", ImGuiTreeNodeFlags_None, "%s", detailed_label.c_str());
    if (detailed_open) {
        for (auto* feature : all_features) {
            if (!feature_matches_filter(feature, filter_buffer)) {
                continue;
            }

            bstring256 feature_label;
            feature_label.printf("%s", feature->name.c_str());

            ImGui::PushID(0x90000000 | ImHashStr(feature_label.c_str()));

            bool feature_open = ImGui::TreeNodeEx(feature_label.c_str(), ImGuiTreeNodeFlags_None, "%s", feature_label.c_str());

            if (feature_open) {
                ImGui::BeginTable(feature_label.c_str(), 2, ImGuiTableFlags_BordersOuter | ImGuiTableFlags_Resizable);

                game_debug_show_property("name", feature->name.c_str());
                if (!feature->text.empty()) {
                    game_debug_show_property("text", feature->text.c_str());
                }
                game_debug_show_property("default", feature->defaultv);

                // Current value based on type
                switch (feature->type()) {
                    case setting_bool: {
                        bool current = feature->to_bool();
                        bool save = current;
                        game_debug_show_property("current_value", current);
                        if (save != current) {
                            feature->set(current);
                        }
                        break;
                    }
                    case setting_string: {
                        xstring current = feature->to_string();
                        game_debug_show_property("current_value", current.c_str());
                        break;
                    }
                    case setting_float:
                    case setting_vec2i:
                    case setting_none:
                        // These types are supported but don't have direct getters in game_feature
                        // The value can be viewed via defaultv or through _settings
                        break;
                }

                // Type information
                bstring256 type_str;
                switch (feature->type()) {
                    case setting_bool:
                        type_str = "bool";
                        break;
                    case setting_float:
                        type_str = "float";
                        break;
                    case setting_vec2i:
                        type_str = "vec2i";
                        break;
                    case setting_string:
                        type_str = "string";
                        break;
                    case setting_none:
                        type_str = "none";
                        break;
                    default:
                        type_str = "unknown";
                        break;
                }
                game_debug_show_property("type", type_str.c_str());

                ImGui::EndTable();
                ImGui::TreePop();
            }

            ImGui::PopID();
        }

        ImGui::TreePop();
    }

    ImGui::TreePop();
}

File: src/scripts/game_features_debug.js
log_info("akhenaten: game_features_debug")

var game_features_debug_filter = ""

function game_features_debug_match_ci(str, substr) {
    if (!substr || substr.length === 0) {
        return true
    }
    if (!str || str.length === 0) {
        return false
    }
    return str.toLowerCase().indexOf(substr.toLowerCase()) >= 0
}

function game_features_debug_matches(name, filter) {
    if (!filter || filter.length === 0) {
        return true
    }
    if (game_features_debug_match_ci(name, filter)) {
        return true
    }
    var text = game_features.text(name)
    if (text && game_features_debug_match_ci(text, filter)) {
        return true
    }
    return false
}

function game_features_debug_draw_filter() {
    imgui.text("Filter:")
    imgui.same_line()
    imgui.set_next_item_width(imgui.content_region_avail_x() - 100)
    game_features_debug_filter = imgui.input_text("##filter", game_features_debug_filter)
    imgui.same_line()
    if (imgui.button("Clear")) {
        game_features_debug_filter = ""
    }
}

function game_features_debug_count_by_type(type_id) {
    var count = 0
    var n = game_features.count
    for (var i = 0; i < n; i++) {
        var name = game_features.name(i)
        if (game_features.type(name) === type_id) {
            count++
        }
    }
    return count
}

function game_features_debug_count_filtered(type_id, filter) {
    var count = 0
    var n = game_features.count
    for (var i = 0; i < n; i++) {
        var name = game_features.name(i)
        if (game_features.type(name) !== type_id) {
            continue
        }
        if (!game_features_debug_matches(name, filter)) {
            continue
        }
        count++
    }
    return count
}

function game_features_debug_count_filtered_all(filter) {
    var count = 0
    var n = game_features.count
    for (var i = 0; i < n; i++) {
        var name = game_features.name(i)
        if (game_features_debug_matches(name, filter)) {
            count++
        }
    }
    return count
}

[es=event_draw_debug_properties]
function game_features_debug_draw_properties(ev) {
    if (!imgui.tree_node_ex("Game Features")) {
        return
    }

    game_features_debug_draw_filter()

    var total_count = game_features.count
    var bool_count = game_features_debug_count_by_type(setting_bool)
    var string_count = game_features_debug_count_by_type(setting_string)
    var float_count = game_features_debug_count_by_type(setting_float)
    var vec2i_count = game_features_debug_count_by_type(setting_vec2i)

    imgui.begin_table("GameFeatures", 2, imgui.table_flags_debug_props())
    imgui.property_input("total_features", total_count)
    imgui.property_input("bool_features", bool_count)
    imgui.property_input("string_features", string_count)
    if (float_count > 0) {
        imgui.property_input("float_features", float_count)
    }
    if (vec2i_count > 0) {
        imgui.property_input("vec2i_features", vec2i_count)
    }
    imgui.end_table()

    var filtered_bool_count = game_features_debug_count_filtered(setting_bool, game_features_debug_filter)
    var filtered_string_count = game_features_debug_count_filtered(setting_string, game_features_debug_filter)

    if (imgui.tree_node_ex2("Bool Features", "Bool Features (" + filtered_bool_count + ")")) {
        imgui.begin_table("BoolFeatures", 2, imgui.table_flags_debug_props())
        var n = game_features.count
        for (var i = 0; i < n; i++) {
            var name = game_features.name(i)
            if (game_features.type(name) !== setting_bool) {
                continue
            }
            if (!game_features_debug_matches(name, game_features_debug_filter)) {
                continue
            }
            __debug_feature_bool(name)
        }
        imgui.end_table()
        imgui.tree_pop()
    }

    if (imgui.tree_node_ex2("String Features", "String Features (" + filtered_string_count + ")")) {
        imgui.begin_table("StringFeatures", 2, imgui.table_flags_debug_props())
        n = game_features.count
        for (i = 0; i < n; i++) {
            name = game_features.name(i)
            if (game_features.type(name) !== setting_string) {
                continue
            }
            if (!game_features_debug_matches(name, game_features_debug_filter)) {
                continue
            }
            imgui.property_input(name, game_features.get(name))
        }
        imgui.end_table()
        imgui.tree_pop()
    }

    var filtered_total = game_features_debug_count_filtered_all(game_features_debug_filter)
    if (imgui.tree_node_ex2("All Features (Detailed)", "All Features (Detailed) (" + filtered_total + ")")) {
        n = game_features.count
        for (i = 0; i < n; i++) {
            name = game_features.name(i)
            if (!game_features_debug_matches(name, game_features_debug_filter)) {
                continue
            }

            if (!imgui.tree_node_ex(name)) {
                continue
            }

            imgui.begin_table(name, 2, imgui.table_flags_debug_props())
            imgui.property_input("name", name)
            var feature_text = game_features.text(name)
            if (feature_text && feature_text.length > 0) {
                imgui.property_input("text", feature_text)
            }

            var default_value = game_features.default(name)
            if (default_value !== null && default_value !== undefined) {
                imgui.property_input("default", default_value)
            }

            var feature_type = game_features.type(name)
            if (feature_type === setting_bool) {
                __debug_feature_bool(name, "current_value")
            } else if (feature_type === setting_string) {
                imgui.property_input("current_value", game_features.get(name))
            }

            imgui.property_input("type", game_features.type_name(feature_type))
            imgui.end_table()
            imgui.tree_pop()
        }
        imgui.tree_pop()
    }

    imgui.tree_pop()
}
